### PR TITLE
feat(cli): add mutation json results

### PR DESCRIPTION
**Pull request by galligan**
## Context

Plan-first and mutating routes need the same stable machine contract as finite read routes. Automation must distinguish previews from writes and see every path actually changed.

## What changed

- adds versioned JSON results to plan-first and mutating routes
- reports explicit `planned` versus `written` state and actual write paths
- preserves `--yes` as write authority; machine output does not imply permission
- reports audit artifacts, release evidence, baselines, generated output, deletions, and acquisitions
- preserves completed work in partial adoption/import failures
- normalizes reported paths and workspace indexing for deterministic automation

## Verification

- required GitHub checks: `changeset`, `check`, and `skillset-ci`
- focused mutation-envelope, adopt/import, reconcile, ledger, and partial-write coverage

## Tracking

Closes SET-288. Stacked on #273.

## Risks / rollout

Partial failures may legitimately return non-empty `writes`; consumers must treat those as completed side effects. Write authority remains explicit and unchanged.

### CODE REVIEW

**Comment by galligan** [comment]
* **#276** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/276?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#275** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/275?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#274** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/274?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/274?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#273** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/273?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#272** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/272?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#271** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/271?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#270** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/270?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#269** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/269?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#268** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/268?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#267** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/267?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#266** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/266?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#265** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/265?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#264** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/264?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#263** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/263?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#262** <a href="https://app.graphite.com/github/pr/outfitter-dev/skillset/262?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: apps/skillset/src/cli-core.ts [hunk 1]
```diff
@@ -406,17 +432,23 @@ export async function runCli(
...
     if (!report.ok || report.blocked) process.exitCode = 1;
     return;
   }
 
   if (command === "restore") {
     if (importPath === undefined) throw new Error("skillset: expected backup id to restore");
     const report = await restoreOutputBackup(rootPath, importPath, { write: yes && !dryRun });
-    printRestoreReport(report);
-    if (!yes || dryRun) console.log("skillset: rerun restore with --yes to write restored files");
+    if (jsonOutput) printCliJsonData("restore", { report, state: report.write ? "written" : "planned", writes: report.restoredPaths });
```
[the remark this answers is not in the document]
Comment (galligan) [reply apps/skillset/src/cli-core.ts]: Fixed: restore previews now report planned state with an empty writes array; only confirmed mutations report restored paths.

File: apps/skillset/src/cli-core.ts [hunk 2]
```diff
@@ -2888,11 +2933,15 @@ function validateJsonFlags(
   route: {
     readonly changeSubcommand?: ChangeSubcommand;
     readonly distributionSubcommand?: DistributionSubcommand;
+    readonly releaseSubcommand?: ReleaseSubcommand;
   }
 ): void {
   if (!jsonOutput) return;
+  if (command === "build" || command === "import" || command === "init" || command === "new" || command === "reconcile" || command === "restore" || command === "update") return;
```
[the remark this answers is not in the document]
Comment (galligan) [reply apps/skillset/src/cli-core.ts:3268]: Fixed: init JSON mode bypasses interactive adoption prompts and always follows the structured result path.

**Comment by galligan** [comment]
@codex review

File: apps/skillset/src/cli-core.ts [hunk 3]
```diff
@@ -253,6 +257,12 @@ export async function runCli(
       return;
     }
     const result = await buildSkillsetResult(rootPath, options);
+    if (jsonOutput) {
+      printCliJsonData("build.apply", { result, state: "written", writes: result.writes.writtenPaths }, result.ok ? 0 : 1, "diagnostics", result.diagnostics);
```
[the remark this answers is not in the document]
Comment (galligan) [reply apps/skillset/src/cli-core.ts]: Fixed: build apply JSON now emits a finite summary without rendered file bytes and reports all actual changed/deleted paths.

File: apps/skillset/src/cli-core.ts [hunk 4]
```diff
@@ -356,8 +373,11 @@ export async function runCli(
         ...changeOptions,
         write: yes && !dryRun,
       });
-      printChangeMigration(report);
-      if ((!yes || dryRun) && report.entries.length > 0) console.log("skillset: rerun change migrate with --yes to rewrite pending entries");
+      if (jsonOutput) printCliJsonData("change.migrate", { report, state: report.written ? "written" : "planned", writes: report.written ? [report.ledgerPath] : [] });
```
[the remark this answers is not in the document]
Comment (galligan) [reply apps/skillset/src/cli-core.ts]: Fixed: change migrate JSON reports migrated destinations, removed source paths, and the ledger only when migration entries actually write.

File: apps/skillset/src/cli-core.ts [hunk 5]
```diff
@@ -625,8 +660,11 @@ export async function runCli(
       useGitRoot: !rootExplicit && importPath === undefined,
       write: yes && !dryRun,
     });
-    printSetupReport(setup, dryRun ? "dry run" : yes ? "written" : "write confirmation required");
-    if (!yes || dryRun) console.log("skillset: rerun init with --yes to write setup files");
+    if (jsonOutput) printCliJsonData("init", { report: setup, state: yes && !dryRun ? "written" : "planned", writes: yes && !dryRun ? setup.files : [] });
```
[the remark this answers is not in the document]
Comment (galligan) [reply apps/skillset/src/cli-core.ts]: Fixed: init JSON now reports created path strings only; existing setup files are excluded from writes.

**Comment by galligan** [comment]
@codex review

File: apps/skillset/src/cli-core.ts [hunk 6]
```diff
@@ -244,9 +243,14 @@ export async function runCli(
   } = parseCliArgs(args);
 
   if (command === "build") {
-    console.log("skillset: build projects source to generated output");
     if (dryRun || !yes) {
       const result = await diffSkillsetResult(rootPath, options);
+      if (jsonOutput) {
+        printCliJsonData("build.plan", { changes: result.data, state: "planned", writes: [] }, result.ok ? 0 : 1, "diagnostics", result.diagnostics);
```
[the remark this answers is not in the document]
Comment (galligan) [reply apps/skillset/src/cli-core.ts]: Fixed: build plan/apply now normalize core diagnostics through serializeDiagnostics. Added a structured-output regression for an unmanaged AGENTS.md collision.

**Comment by galligan** [comment]
@codex review

File: apps/skillset/src/cli-core.ts [hunk 7]
```diff
@@ -681,8 +745,11 @@ export async function runCli(
       skillsetOptions: options,
       write: yes && !dryRun,
     });
-    printNewSourceReport(report, dryRun ? "dry run" : yes ? "written" : "write confirmation required");
-    if (!yes || dryRun) console.log("skillset: rerun new with --yes to write source files");
+    if (jsonOutput) printCliJsonData("new", { report, state: report.write ? "written" : "planned", writes: report.write ? report.files : [] });
```
[the remark this answers is not in the document]
Comment (galligan) [reply apps/skillset/src/cli-core.ts]: Fixed in the PR: new command JSON now maps created file records to string paths, with a regression test asserting a uniform string array.

File: apps/skillset/src/cli-core.ts [hunk 8]
```diff
@@ -301,23 +321,27 @@ export async function runCli(
...
-      printChangeEntry("added", (await addChangeEntry(rootPath, {
+      const report = await addChangeEntry(rootPath, {
         ...changeOptions,
         ...(changeBump === undefined ? {} : { bump: changeBump }),
         ...(changeGroup === undefined ? {} : { group: changeGroup }),
         reason: changeReason ?? { kind: "auto" },
         scopes: changeScopes ?? [],
-      })).entry);
+      });
+      if (jsonOutput) printCliJsonData("change.add", { report, state: "written", writes: [report.entry.path] });
```
[the remark this answers is not in the document]
Comment (galligan) [reply apps/skillset/src/cli-core.ts]: Fixed in the PR: change add now reports both the pending entry and ledger JSONL path in writes; the regression test asserts the exact pair.

**Comment by galligan** [comment]
@codex review

File: apps/skillset/src/cli-core.ts [hunk 9]
```diff
@@ -306,23 +326,27 @@ export async function runCli(
...
       if (changeRef === undefined) throw new Error("skillset: change reason requires @ref");
-      printChangeEntry("updated", (await updateChangeReason(rootPath, {
+      const report = await updateChangeReason(rootPath, {
         ...changeOptions,
         append: changeAppend,
         reason: changeReason ?? { kind: "auto" },
         ref: changeRef,
-      })).entry);
+      });
+      if (jsonOutput) printCliJsonData("change.reason", { report, state: "written", writes: [report.entry.path] });
```
[the remark this answers is not in the document]
Comment (galligan) [reply apps/skillset/src/cli-core.ts]: Fixed in the PR: current reason-only entries report ledgerPath and change reason JSON includes both the entry and ledger in writes; legacy frontmatter entries remain accurately conditional.

File: apps/skillset/src/cli-core.ts [hunk 10]
```diff
@@ -306,23 +326,27 @@ export async function runCli(
...
-      printChangeEntry("added", (await addChangeEntry(rootPath, {
+      const report = await addChangeEntry(rootPath, {
         ...changeOptions,
         ...(changeBump === undefined ? {} : { bump: changeBump }),
         ...(changeGroup === undefined ? {} : { group: changeGroup }),
         reason: changeReason ?? { kind: "auto" },
         scopes: changeScopes ?? [],
-      })).entry);
+      });
+      if (jsonOutput) printCliJsonData("change.add", { report, state: "written", writes: [report.entry.path, report.ledgerPath] });
```
[the remark this answers is not in the document]
Comment (galligan) [reply apps/skillset/src/cli-core.ts]: Fixed in the PR: add, reason, and amend mutation JSON now serialize ChangeEntryView sourceHashes through the same serializer used by read routes; the structured add test asserts non-empty hash evidence.

**Comment by galligan** [comment]
@codex review

File: apps/skillset/src/cli-core.ts [hunk 11]
```diff
@@ -650,8 +721,11 @@ export async function runCli(
       useGitRoot: !rootExplicit && importPath === undefined,
       write: yes && !dryRun,
     });
-    printSetupReport(setup, dryRun ? "dry run" : yes ? "written" : "write confirmation required");
-    if (!yes || dryRun) console.log("skillset: rerun init with --yes to write setup files");
+    if (jsonOutput) printCliJsonData("init", { report: setup, state: yes && !dryRun ? "written" : "planned", writes: yes && !dryRun ? setup.files.filter((file) => file.status === "create").map((file) => file.path) : [] });
```
[the remark this answers is not in the document]
Comment (galligan) [reply apps/skillset/src/cli-core.ts]: Fixed: init reports the release-state path whenever baseline seeding actually writes it, and leaves no-op writes empty.

File: apps/skillset/src/cli-core.ts [hunk 12]
```diff
@@ -665,7 +739,9 @@ export async function runCli(
       rootPath,
       ...(options.sourceDir === undefined ? {} : { sourceDir: options.sourceDir }),
     });
-    if (result.imports.length === 1) {
+    if (jsonOutput) {
+      printCliJsonData("import", { result, state: "written", writes: result.imports.flatMap((entry) => [entry.targetPath]) });
```
[the remark this answers is not in the document]
Comment (galligan) [reply apps/skillset/src/cli-core.ts]: Fixed: import propagates baseline-state writes into the JSON mutation audit and deduplicates paths across batch imports.

**Comment by galligan** [comment]
@codex review

File: apps/skillset/src/cli-core.ts [hunk 13]
```diff
@@ -593,13 +661,16 @@ export async function runCli(
         write: writeMode,
       });
       const reason = writeMode ? report.write ? "written" : "blocked before write" : "write confirmation required";
-      printAdoptReport(report, reason);
-      if (!writeMode && report.ok && initAdopt !== undefined) console.log("skillset: rerun init with --adopt and --yes to write adopted source");
+      if (jsonOutput) printCliJsonData("init.adopt", { report, state: report.write ? "written" : "planned", writes: report.write ? [...report.setupFiles.filter((file) => file.status === "create").map((file) => file.path), ...report.imports.flatMap((entry) => entry.destination === undefined ? [] : [entry.destination])] : [] }, report.ok ? 0 : 1);
```
[the remark this answers is not in the document]
Comment (galligan) [reply apps/skillset/src/cli-core.ts]: Fixed in 956b45c. Init-adoption JSON now reports setup files, instruction destinations, adopted skill/plugin units, and baseline state with stable deduplication.

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
@codex review

**Comment by galligan** [comment]
### Merge activity

* **Jul 14, 2:45 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/outfitter-dev/skillset/274).
* **Jul 14, 3:09 PM UTC**: [Graphite](https://app.graphite.com/github/pr/outfitter-dev/skillset/274) rebased this pull request as part of a merge.
* **Jul 14, 3:09 PM UTC**: @galligan merged this pull request with [Graphite](https://app.graphite.com/github/pr/outfitter-dev/skillset/274).
